[BEAM-5187] Add a ProcessJobBundleFactory for process-based execution

mxm opened the pull request:
This adds a ProcessJobBundleFactory for process-based execution which currently
takes the Environment URL String and executes it with the regular parameters
which are part of the container contract.

As of now, the factory needs to be wired manually. This will change when the
Runner API's Environment supports process-based execution.

The next steps are:

- Modifying the Environment Proto to explicitly support process-based execution
  alongside with other parameters (e.g. target platform, OS).

- Introduce artifact staging

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

mxm commented:
CC @tweise @angoenka

tweise commented:
@mxm thanks for taking this up!

I used the changes in this PR to illustrate what I would like to accomplish as custom extension here: https://github.com/tweise/beam/commits/processJobBundleFactory

My goal is to launch the Python SDK harness directly (eliminating boot.go), artifact staging can be skipped.

From cursory test it seems error handling in the process launch needs some fixing. We also need the ability to propagate the environment of the JVM plus set additional environment variables to meet the contract of the Python worker.

angoenka commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/JobBundleFactoryBase.java:
```diff
@@ -0,0 +1,331 @@
+ */
+@ThreadSafe
+public abstract class JobBundleFactoryBase implements JobBundleFactory {
+  protected final Logger logger = LoggerFactory.getLogger(getClass());
```
lets use LOG

mxm replied:
The logger is now non-static and only in the base class, that's why the variable name is lower case. Can create static loggers in both classes if you prefer that.

angoenka commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/ProcessEnvironmentFactory.java:
```diff
@@ -0,0 +1,154 @@
+            String.format("--provision_endpoint=%s", provisionEndpoint),
+            String.format("--control_endpoint=%s", controlEndpoint));
+
+    LOG.debug("Creating Process with ID {}", workerId);
```
`Creating Process with ID {}` -> `Creating Process for Worker ID {}`

angoenka commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/ProcessManager.java:
```diff
@@ -0,0 +1,83 @@
+
+  private final Map<String, Process> processes;
+
+  public static ProcessManager getDefault() {
```
`getDefault ` -> `getInstance` as this is the only way to get an instance.

mxm commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DockerJobBundleFactory.java:
```diff
@@ -234,7 +125,7 @@ protected ServerFactory getServerFactory() {
             () -> MAC_PORT.getAndUpdate(val -> val == MAC_PORT_END ? MAC_PORT_START : val + 1));
       default:
-        LOG.warn("Unknown Docker platform. Falling back to default server factory");
+        logger.warn("Unknown Docker platform. Falling back to default server factory");
```
The logger is now non-static and only in the base class, that's why the variable name is lower case. Can create static loggers in both classes if you prefer that.

tweise replied:
I think there should be multiple instances since we want to see where the logging actually occurred (superclass vs. subclass).

mxm replied:
Makes sense. Re-introduced the static instances.

mxm commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/ProcessEnvironmentFactory.java:
```diff
@@ -0,0 +1,154 @@
+    this.clientSource = clientSource;
+  }
+
+  /** Creates a new, active {@link RemoteEnvironment} backed by a forked process. */
```
The artifact retrieval is not yet implemented. It's still up to debate, whether we reuse the `boot.go` code or implement this in another way. Until it is implemented the user has to ship libraries manually.

I would prefer `boot.go`. It takes a compile-time dependency on Go but we could cross-compile it for various platforms.

tweise replied:
The alternative would be rewriting the boot.go logic in Java and include it into ProcessEnvironmentFactory.

tweise replied:
It would remove the added deployment complexity for users.

mxm commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/JobBundleFactoryBase.java:
```diff
@@ -0,0 +1,331 @@
+ */
+@ThreadSafe
+public abstract class JobBundleFactoryBase implements JobBundleFactory {
+  protected final Logger logger = LoggerFactory.getLogger(getClass());
```
That's already the case because `getClass()` always returns the class name of the top-level class.

mxm commented:
Thanks for checking it out @angoenka @tweise.

I've checked out your code @tweise.

>From cursory test it seems error handling in the process launch needs some fixing. 

The code actually fails if the binary path is not correct. It doesn't fail immediately when the process fails but it will time out. I'll introduce a check for the exit code when waiting for the connection from the SDK harness process.
 
>We also need the ability to propagate the environment of the JVM plus set additional environment variables to meet the contract of the Python worker.

As things stand now, setting the correct environment variables is the responsibility of your startup script. But you're right, it doesn't make sense to pass the container contract parameters because the SDK harness processes rely on environment variables. Since we launch the process directly, we can set the required environment variables for the SDK harness. 

What do you want to achieve by propagating the environment of the JVM?

mxm commented:
@tweise I've improve the process startup. It will now fail immediately if the process exists prematurely. Considering the container contract variables, I don't think it makes sense to set them as environment variables. This works for the Python SDK but not for Go because Go does not use environment variables. We could change that but I'm not sure if that is a good idea because that would make a double contract. Instead, I think your startup executable should decide what to do with the provided parameters, e.g. set environment variables before bringing up the SDK harness.

tweise commented:
I was able to launch the python process directly when running the JVM within virtualenv. Environment variable are forwarded by default. Just need the ability to specify additional variables that are needed by the worker (in my customized launcher).

mxm commented:
>It may actually be good to also capture the output.

For debugging purposes we could save some of the output but I don't want to read all the output because it could create memory pressure. For the startup the output might be helpful but everything else should go through the logging server.

>As  for the environment variables, my suggestion wasn't to use environment variables in your implementation, just support the ability to specify  them in runCommand since it is one of the basic ingredients of launching process and makes it reusable.

I see, that can be done.

>I was able to launch the python process directly when running the JVM within virtualenv.

Great :)

tweise commented:
Probably all of the process stdout/stderr output should go to runner logging, irrespective of whether at startup or later? If we redirect it elsewhere, then important information may be missing, such as when there is an issue with process startup or unexpected termination.

tweise commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/ProcessManager.java:
```diff
@@ -0,0 +1,157 @@
+
+    LOG.debug("Attempting to start process with command: " + pb.command());
+    // Pipe stdout and stderr to /dev/null to avoid blocking the process due to filled PIPE buffer
+    pb.redirectErrorStream(true);
```
Consider inheriting stderr (and possibly also stdout), possibly optional, since it may be hard to diagnose issues when the output isn't available.

tweise commented:
@mxm with a customization I got the python worker to come up without the boot.go (https://github.com/tweise/beam/commit/006ad96de84a839f8f07abca70c67f5549eca0a1)

Please see the diff, there are a few changes that would be needed in your PR to support it (primarily modifiers).

btw I noticed that after job server shutdown, launched processes still stick around and don't exit

tweise commented:
Something is wrong with the bookkeeping of processes in `ProcessManager`.

If I repeat a pipeline run before the environment was cleaned up (which would happen after 30s?), the process manager will find that a process with the id already exists, and fail.

```
Caused by: java.lang.IllegalStateException: There was already a process running with id 1
        at org.apache.beam.runners.fnexecution.environment.ProcessManager.runCommand(ProcessManager.java:78)
```

angoenka commented:
The problem might be with static ProcessManager.getInstance.
Should we create new instance of ProcessManager for each JobInfo as we do in DockerJobBundleFactory

angoenka commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/ProcessManager.java:
```diff
@@ -0,0 +1,157 @@
+  private final Map<String, Process> processes;
+
+  public static ProcessManager getInstance() {
+    return INSTANCE;
```
Having a singleton here means we have a single map of processes across jobs.

mxm replied:
Absolutely, the `ProcessManager` needs to be instantiated per `ProcessEnvironmentFactory`. Getting rid of the static instance also makes testing easier.

mxm commented:
> btw I noticed that after job server shutdown, launched processes still stick around and don't exit

Please see the latest version of the PR from yesterday. I revised the shutdown logic to eventually kill  processes if they don't stop gracefully. In your code you're using the old version. Also, I've added a ShutdownHook to kill running processes in case the JVM shuts down prematurely.

@tweise @angoenka I've removed the singleton ProcessManager and instantiate it per `ProcessEnvironmentFactory` which should get rid of the duplicate worker ids you were seeing @tweise.

@tweise For the IO, I've enabled inheriting IO when the log level is set to DEBUG. That should help with debugging process startup problems.

tweise commented:
@mxm rebased my branch https://github.com/tweise/beam/commit/c64add1a4eb2e1a4ae818e7891516a5c57ef1fe1

Can you add the changes to ProcessManager and DockerJobBundleFactory to your PR?

mxm commented:
@tweise sure, done.

tweise commented:
thanks!

One more observation: The shutdown hook kills the top level process, but not its child process(es). For  bash -> python, only bash will be killed. Maybe we can add a (very small) graceful termination period before taking out the hammer?

mxm commented:
@tweise I made the ShutdownHook gracefully stop the processes, followed by a kill if there are still running processes.

tweise commented:
Please run:  `./gradlew :beam-runners-java-fn-execution:check`

mxm commented:
@tweise Fixed.
